File: sandbox-runner/patterns/local-activities/python/activities.py
import time
from temporalio import activity
from shared import Transaction, TransactionRequest


@activity.defn
async def validate_transaction(req: TransactionRequest) -> Transaction:
    if req.amount <= 0:
        raise ValueError(f"Invalid amount: {req.amount}")
    tx_id = f"tx-{int(time.time() * 1000)}"
    return Transaction(id=tx_id, status="initialized")


@activity.defn
async def reserve_funds(tx: Transaction) -> Transaction:
    return Transaction(id=tx.id, status="reserved")


@activity.defn
async def settle_transaction(tx: Transaction) -> Transaction:
    return Transaction(id=tx.id, status="completed")

File: sandbox-runner/patterns/local-activities/typescript/activities.ts
import { TransactionRequest, Transaction } from "./shared";

export async function validateTransaction(req: TransactionRequest): Promise<Transaction> {
  console.log(`Validating ${req.amount} ${req.currency}`);
  if (req.amount <= 0) throw new Error(`Invalid amount: ${req.amount}`);
  return { id: `tx-${Date.now()}`, status: "initialized" };
}

export async function reserveFunds(tx: Transaction): Promise<Transaction> {
  console.log(`Reserving funds for transaction ${tx.id}`);
  return { id: tx.id, status: "reserved" };
}

export async function settleTransaction(tx: Transaction): Promise<Transaction> {
  console.log(`Settling transaction ${tx.id}`);
  return { id: tx.id, status: "completed" };
}
